### PR TITLE
Getting Interface.load() working for 2.x and 3.x models and Spaces

### PR DESCRIPTION
**Pull request by abidlabs**
Took waaay longer than I thought, but there were a lot of cases to cover. We can now load any model, 2.x space, or 3.x space using `gr.Interface.load()` or basically its alias `gr.Blocks.load()`. The basic idea here is:

- When loading a `model`, it creates an `Interface` based on the type of `model`
- When loading a 2.x Space, it creates an `Interface` based on the config (very easy mapping as the config is just basically the keyword arguments needed to instantiate the `Interface`)
- When loading a 3.x Space, it reads the config and recreates the Blocks and components locally. Of course the backend functions are a little different -- instead of calling the local function, it instead calls the remote /api/predict appropriately

Touches a lot of the backend code, so happy to explain anything

You can test in different scenarios:

- Loading a **model** from the Hub

```py
gr.Interface.load("models/gpt2", title="GPT 2", description="a completion model").launch()
```

```py
gr.Interface.load("models/facebook/wav2vec2-base-960h", inputs="mic").launch()
```

- Loading a **Gradio 2.x Space** from the Hub

```py
gr.Interface.load("spaces/abidlabs/test5").launch()
```

- Loading a **Gradio 3.x Space** from the Hub

```py
gr.Interface.load("spaces/gradio/pictionary").launch()
```

- Loading a **Gradio 3.x Space** from the Hub within a Blocks

```py
with gr.Blocks() as demo:
    with gr.Tabs():
        with gr.TabItem("Pictionary"):
            gr.Interface.load("spaces/gradio/pictionary")

demo.launch()
```

- Creating a Blocks that loads both a **Gradio 3.x Space** and a **Gradio 2.x Space**

```py
with gr.Blocks() as demo:
    with gr.Tabs():
        with gr.TabItem("Pictionary"):
            gr.Interface.load("spaces/gradio/pictionary")
        with gr.TabItem("Hello World"):
            gr.Interface.load("spaces/abidlabs/test5")

demo.launch()
```

 
Closes: #1248 cc @apolinario

### CODE REVIEW

**Comment by FarukOzderim** [comment]
Can we load Blocks spaces as well?

**Comment by abidlabs** [comment]
Yup! It should work

**Comment by apolinario** [comment]
This is looking amazing! 
I have installed it locally and this was my quick experience:

- Loading a Gradio 2.x Space from the Hub ✅
`latent = gr.Interface.load("spaces/multimodalart/latentdiffusion")` 
This is working

- Loading a Gradio 3.x Space from the Hub ❌ 
`dalle = gr.Interface.load('spaces/dalle-mini/dalle-mini')`
`pictionary = gr.Interface.load("spaces/gradio/pictionary")`
This didn't work here. Error below. Do the host spaces have to be updated with this upgrade too for it to work?
```
Fetching interface from: https://huggingface.co/spaces/dalle-mini/dalle-mini
Traceback (most recent call last):
  File "/mnt/c/Users/poli/Documents/GitHub/gradio/app.py", line 3, in <module>
    dalle = gr.Interface.load('spaces/dalle-mini/dalle-mini')
  File "/mnt/c/Users/poli/Documents/GitHub/gradio/gradio/interface.py", line 90, in load
    return super().load(name=name, src=src, api_key=api_key, alias=alias, **kwargs)
  File "/mnt/c/Users/poli/Documents/GitHub/gradio/gradio/blocks.py", line 515, in load
    return external.load_blocks_from_repo(name, src, api_key, alias, **kwargs)
  File "/mnt/c/Users/poli/Documents/GitHub/gradio/gradio/external.py", line 38, in load_blocks_from_repo
    blocks: gradio.Blocks = factory_methods[src](name, api_key, alias, **kwargs)
  File "/mnt/c/Users/poli/Documents/GitHub/gradio/gradio/external.py", line 313, in get_spaces
    return get_spaces_blocks(model_name, config)
  File "/mnt/c/Users/poli/Documents/GitHub/gradio/gradio/external.py", line 350, in get_spaces_blocks
    return gradio.Blocks.from_config(config, fns)
  File "/mnt/c/Users/poli/Documents/GitHub/gradio/gradio/blocks.py", line 307, in from_config
    event_method(fn=fn, **dependency)
  File "/mnt/c/Users/poli/Documents/GitHub/gradio/gradio/events.py", line 68, in click
    self.set_event_trigger(
  File "/mnt/c/Users/poli/Documents/GitHub/gradio/gradio/blocks.py", line 115, in set_event_trigger
    Context.root_block.fns.append(BlockFunction(fn, preprocess, postprocess))
AttributeError: 'NoneType' object has no attribute 'fns'
``` 

Also, conceptually, about this
> When loading a 3.x Space, it reads the config and recreates the Blocks and components locally. Of course the backend functions are a little different -- instead of calling the local function, it instead calls the remote /api/predict appropriately
I may have misunderstood, but with this would I lose the loading 2.x blocks API functionality? 
I really like the way it loads Interfaces, like: 
``` 
inputs:
|-textbox
|-slider
|-radio
|-radio
|-slider
|-slider
outputs:
|-image
|-carousel
|-textbox
```
When I load a 3.x Spaces I don't know if necessarily I want to use the blocks of the loaded space as actual blocks on the new space. It may be the case I still want to use them as an API like the 2.x blocks. 

I have an idea on how this could work: it could check each `button.click` component there and create a mock API for each of the buttons
For example, MindsEye Lite has this: 
`get_image_latent.click(text2image_latent, inputs=[text,steps,width,height,images,diversity], outputs=gallery)`
`get_image_rudalle.click(text2image_rudalle, inputs=[text,aspect,model], outputs=gallery)`
So the response could be two APIs on this format: 
``` 
get_image_latent
inputs:
|-textbox
|-slider
|-slider
|-slider
|-slider
outputs:
|-gallery
```

``` 
get_image_rudalle
inputs:
|-textbox
|-radio
|-dropdown
outputs:
|-gallery
```
(PS: As I couldn't test, if this is already possible feel free to ignore)

**Comment by apolinario** [comment]
Also, conceptually, about this
> When loading a 3.x Space, it reads the config and recreates the Blocks and components locally. Of course the backend functions are a little different -- instead of calling the local function, it instead calls the remote /api/predict appropriately

I may have misunderstood, but with this would I lose the loading 2.x blocks API functionality? 
I really like the way it loads Interfaces for 2.x, like: 
``` 
inputs:
|-textbox
|-slider
|-radio
|-radio
|-slider
|-slider
outputs:
|-image
|-carousel
|-textbox
```
When I load a 3.x Spaces I don't know if necessarily I want to use the blocks of the loaded space as actual blocks on the new space. It may be the case I still want to use them as an API like in the 2.x blocks

I have an idea on how this could work: it could check each `button.click` component on the 3.x block it's trying to load and create this 'API' for each of the buttons. For example, MindsEye Lite has this: 
`get_image_latent.click(text2image_latent, inputs=[text,steps,width,height,images,diversity], outputs=gallery)`
`get_image_rudalle.click(text2image_rudalle, inputs=[text,aspect,model], outputs=gallery)`
So if someone wanted to load MindsEye Lite onto their Spaces - besides being able to use MindsEye Lite blocks (which I think is neat) - the response could be two APIs on this format: 
``` 
get_image_latent
inputs:
|-textbox
|-slider
|-slider
|-slider
|-slider
outputs:
|-gallery
```

``` 
get_image_rudalle
inputs:
|-textbox
|-radio
|-dropdown
outputs:
|-gallery
```
(PS: As I couldn't test, if this is already possible feel free to ignore)

File: gradio/external.py [hunk 1]
```diff
@@ -1,13 +1,45 @@
...
 import re
+from copy import deepcopy
+from typing import Callable, Dict
 
 import requests
 
+import gradio
 from gradio import components, utils
 
+####################################################
```
Comment (aliabid94) [inline gradio/external.py:15]: Remove this comment.

**Comment by aliabid94** [comment]
LGTM

**Comment by abidlabs** [comment]
Thanks @apolinario can confirm the issue. Working on a fix!

**Comment by abidlabs** [comment]
> I may have misunderstood, but with this I wouldn't have the 2.x-like functions with that I can call to invoke the results from the Space? I really like the way it loads Interfaces for 2.x, like:

Yes, you are right, this wouldn't currently be possible. However, we are working on this as part of bringing back the API endpoints for Blocks

**Comment by abidlabs** [comment]
> Loading a Gradio 3.x Space from the Hub ❌
> dalle = gr.Interface.load('spaces/dalle-mini/dalle-mini')
> pictionary = gr.Interface.load("spaces/gradio/pictionary")
> This didn't work here. Error below. Do the host spaces have to be updated with this upgrade too for it to work?

Nope host spaces don't need to be updated, that was just a bug on my part. Fixed now, I tested both and they are working!

Let me know if something doesn't work for you and I'll squash the bug, but merging this so you can go ahead and try it out more easily

**Comment by apolinario** [comment]
Feel free to check out the `gr.Interface.load()` use-case I have for MindsEye Lite. I do not use the `launch()` for them: https://huggingface.co/spaces/multimodalart/mindseye-lite/blob/main/app.py

I feel the simplicity of this is great: 
```
vqgan = gr.Interface.load("spaces/multimodalart/vqgan")
(...)
def text2image_vqgan(text,width,height,style,steps,flavor):
    results = vqgan(text,width,height,style,steps,flavor)
    return([results])
(...)
get_image_vqgan.click(text2image_vqgan, inputs=[text,width_vq,height_vq,style,steps_vq,flavor],outputs=gallery)
```

**Pull request merged**